### PR TITLE
Overhaul 2 - New features and complete functionality

### PR DESCRIPTION
**Pull request by VarLad**
This PR includes support for 
1. Better multifunction encoding
2. Option to choose from different encoding techniques
3. Evolution enhanced mapping
4. Parameterisation of cost function and optimized boundary handling
5. Regularization and boundary loss for the above
 
Other than that, better tests which aim to catch a wrong function call, incorrect function definition and other such problems through  all the features.

### CODE REVIEW

File: src/QuantumNLDiffEq.jl [hunk 1]
```diff
@@ -1,64 +1,137 @@
...
+
+function rrule(func::ODEFunction, u, p, t)
+    y = func(u,p,t)
+    function func_pullback(ȳ)
+        Ju = jacobian((u)->func(u,p,t),u)
+        Jp = jacobian((p)->func(u,p,t),p)
+        NoTangent(), Ju'*ȳ, Jp'*ȳ, NoTangent()
+    end
+    return y, func_pullback
+end
```
Comment (ChrisRackauckas) [inline src/QuantumNLDiffEq.jl:18]: This kind of piracy is dangerous. Instead use a wrapper function and then define this on the wrapper.

Comment (VarLad) [reply src/QuantumNLDiffEq.jl:18]: I guess `func::ODEFunction` is dangerous?
Thanks, will fix that!

**Comment by VarLad** [comment]
Pasting a png file here
jl_SbFEaMjEmj

File: src/QuantumNLDiffEq.jl (hunk 1 above)
Comment (VarLad) [reply src/QuantumNLDiffEq.jl]: @ChrisRackauckas Does the latest commit fix that?

**Comment by ChrisRackauckas** [comment]
Looks great!

Make an issue about improving that derivative handling, but it's more than fine enough for now.

**Pull request merged**